datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,412 @@
+    ///                 (Expr::Literal(left), Expr::Literal(right)) => {
+    ///                     Ok((self.f)(left, right)?)
+    ///                 }
+    ///                 _ => Ok(Transformed::no(node)),
+    ///             },
+    ///             _ => Ok(Transformed::no(node)),
+    ///         }
+    ///     }
+    /// }
+    /// // create two rewriters for evaluating literals
+    /// // first rewriter evaluates addition expressions
+    /// let mut addition_rewriter = ConstBinaryExprRewriter {
+    ///     op: Operator::Plus,
+    ///     f: Box::new(|left, right| {
+    ///         Ok(Transformed::yes(Expr::Literal(left.add(right)?)))
+    ///     }),
+    /// };
+    /// // second rewriter evaluates multiplication expression
+    /// let mut multiplication_rewriter = ConstBinaryExprRewriter {
+    ///     op: Operator::Multiply,
+    ///     f: Box::new(|left, right| {
+    ///         Ok(Transformed::yes(Expr::Literal(left.mul(right)?)))
+    ///     }),
+    /// };
+    /// // Create an expression from constant literals
+    /// let expr = lit(6) + (lit(4) * (lit(2) + (lit(3) * lit(5))));
+    /// // Run rewriters in a loop until constant expression is fully evaluated
+    /// let (evaluated_expr, info) = RewriteCycle::new()
+    ///     .with_max_cycles(4)
+    ///     .each_cycle(expr, |cycle_state| {
+    ///         cycle_state
+    ///             .rewrite(&mut addition_rewriter)?
+    ///             .rewrite(&mut multiplication_rewriter)
+    ///     })
+    ///     .unwrap();
+    /// assert_eq!(evaluated_expr, lit(74));
+    /// assert_eq!(info.completed_cycles(), 3);
+    /// assert_eq!(info.total_iterations(), 7);
+    /// ```
+    pub fn each_cycle<
+        Node: TreeNode,
+        F: FnMut(
+            RewriteCycleState<Node>,
+        ) -> RewriteCycleControlFlow<RewriteCycleState<Node>>,
+    >(
+        &self,
+        node: Node,
+        mut f: F,
+    ) -> Result<(Node, RewriteCycleInfo)> {
+        let mut state = RewriteCycleState::new(node);
+        if self.max_cycles == 0 {
+            return state.finish();
+        }
+        // run first cycle then record number of rewriters
+        state = match f(state) {
+            ControlFlow::Break(result) => return result?.finish(),
+            ControlFlow::Continue(node) => node,
+        };
+        state.record_cycle_length();
+        if state.is_done() {
+            return state.finish();
+        }
+        // run remaining cycles
+        match (1..self.max_cycles).try_fold(state, |state, _| f(state)) {
+            ControlFlow::Break(result) => result?.finish(),
+            ControlFlow::Continue(state) => state.finish(),
+        }
+    }
+}
+
+/// Iteration state of a rewrite cycle. See [RewriteCycle::each_cycle] for usage examples and information.
+#[derive(Debug)]
+pub struct RewriteCycleState<Node: TreeNode> {
+    node: Node,
+    consecutive_unchanged_count: usize,
```

Comment:
I think it would help me follow the code better if these fields used the terms "cycle" and "iteration" consistently

Like

```rust
    consecutive_unchanged_iterations: usize,
    iteration_count: usize,
    // the number of iterations in each cycle (set after the first cycle)
    cycle_length: Option<usize>,
```